services/director-v2/tests/unit/test_dynamic_sidecar_monitor.py

```diff
@@ -0,0 +1,408 @@
+from typing import AsyncGenerator, Callable, Iterator, List, Type
+from unittest.mock import AsyncMock
+
+import httpx
+import pytest
+import respx
+from _pytest.monkeypatch import MonkeyPatch
+from fastapi import FastAPI
+from pytest_mock.plugin import MockerFixture
+from respx.router import MockRouter
+from simcore_service_director_v2.core.settings import (
+    AppSettings,
+    DynamicServicesSettings,
+)
+from simcore_service_director_v2.models.schemas.dynamic_services import (
+    DynamicSidecarStatus,
+    MonitorData,
+    RunningDynamicServiceDetails,
+    ServiceState,
+)
+from simcore_service_director_v2.modules.dynamic_sidecar.client_api import (
+    get_url,
+    setup_api_client,
+    shutdown_api_client,
+)
+from simcore_service_director_v2.modules.dynamic_sidecar.errors import (
+    DynamicSidecarError,
+    DynamicSidecarNotFoundError,
+)
+from simcore_service_director_v2.modules.dynamic_sidecar.monitor import (
+    DynamicSidecarsMonitor,
+    setup_monitor,
+    shutdown_monitor,
+    task,
+)
+from simcore_service_director_v2.modules.dynamic_sidecar.monitor.events import (
+    REGISTERED_EVENTS,
+    MonitorEvent,
+)
+
+# running monitor at a hight rate to stress out the system
+# and ensure faster tests
+TEST_MONITOR_INTERVAL_SECONDS = 0.01
+SLEEP_TO_AWAIT_MONITOR_TRIGGERS = 10 * TEST_MONITOR_INTERVAL_SECONDS
+
+pytestmark = pytest.mark.asyncio
+
+log = logging.getLogger(__name__)
+
+
+# UTILS
+
+
+@contextmanager
+def _mock_containers_docker_status(
+    monitor_data: MonitorData, expected_response: httpx.Response
+) -> Iterator[MockRouter]:
+    service_endpoint = monitor_data.dynamic_sidecar.endpoint
+    with respx.mock as mock:
+        mock.get(
+            re.compile(
+                fr"^http://{monitor_data.service_name}:{monitor_data.dynamic_sidecar.port}/v1/containers\?only_status=true"
+            ),
+            name="containers_docker_status",
+        ).mock(return_value=expected_response)
+
+        mock.post(
+            get_url(service_endpoint, "/v1/containers:down"),
+            name="begin_service_destruction",
+        ).respond(text="")
+
+        yield mock
+
+
+@asynccontextmanager
+async def _assert_get_dynamic_services_mocked(
+    monitor: DynamicSidecarsMonitor,
+    monitor_data: MonitorData,
+    mock_service_running: AsyncMock,
+    expected_response: httpx.Response,
+) -> AsyncGenerator[RunningDynamicServiceDetails, None]:
+    with _mock_containers_docker_status(monitor_data, expected_response):
+        await monitor.add_service_to_monitor(monitor_data)
+
+        stack_status = await monitor.get_stack_status(monitor_data.node_uuid)
+        assert mock_service_running.called
+
+        yield stack_status
+
+        await monitor.remove_service_from_monitor(monitor_data.node_uuid, True)
+
+
+# FIXTURES
+
+
+@pytest.fixture(scope="module", autouse=True)
+def ensure_in_docker_swarm(docker_swarm: None) -> None:
+    pass
+
+
+@pytest.fixture(scope="module", autouse=True)
```

Comment:
the day you have to debug a test you will not like these autouse ... :-)

Reply:
OK, OK, I got it, you want me to change them 🤣